### PR TITLE
Cleanup state of the builders when build is done.

### PR DESCRIPTION
**Pull request by AGulev**
Currently, some builders save intermediate states to reduce work between different "create" and "build" stages. This may cause additional memory pressure on large projects, so it's better to release memory and allow the garbage collector to free it.

### CODE REVIEW

**Comment by AGulev** [comment]
It's a draft because I want to take a look what can be improved in time profiler so it consumes less memory

**Comment by AGulev** [comment]
I'll make a separate PR for time profiler if I find sometihng

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 1]
```diff
@@ -72,7 +72,7 @@ public abstract class LuaBuilder extends Builder {
     private static List<ILuaPreprocessor> luaPreprocessors = null;
     private static List<ILuaObfuscator> luaObfuscators = null;
 
-    private Map<String, LuaScanner> luaScanners = new HashMap();
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:75]: Each file has it's own builder. So one file has only one Lua Scanner. There is no reason to have more than on LuaScanner

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java [hunk 2]
```diff
@@ -156,4 +156,10 @@ public void build(Task task) throws CompileExceptionError,
         task.output(0).setContent(out.toByteArray());
     }
 
+    @Override
+    public void clearState() {
+        super.clearState();
+        protoParams = null;
+        srcBuilders = null;
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/ProtoBuilder.java:163]: cleanup parsed data

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 3]
```diff
@@ -1881,6 +1881,7 @@ else if (!resourceCache.contains(key)) {
                             state.putSignature(r.getAbsPath(), taskSignature);
                         }
                     }
+                    builder.clearState();
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:1884]: when task build is done - cleanup builder state

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java [hunk 4]
```diff
@@ -461,4 +461,9 @@ protected CollectionDesc.Builder transform(Task task, IResource resource, Collec
         return messageBuilder;
     }
 
+    @Override
+    public void clearState() {
+        super.clearState();
+        compCounterInputsCount = null;
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/CollectionBuilder.java:467]: Cleanup data used to count components

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java [hunk 5]
```diff
@@ -623,4 +621,10 @@ private PropertyDeclarations buildProperties(IResource resource, List<LuaScanner
         }
         return builder.build();
     }
+
+    @Override
+    public void clearState() {
+        super.clearState();
+        luaScanner = null;
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/LuaBuilder.java:628]: cleanup scanner

File: com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java [hunk 6]
```diff
@@ -1790,9 +1790,6 @@ private List<TaskResult> runTasks(IProgress monitor) throws IOException, Compile
 
                 // compare all task signature. current task signature between previous
                 // signature from state on disk
-                TimeProfiler.start("compare signatures");
-                TimeProfiler.addData("color", "#FFC0CB");
-                TimeProfiler.addData("main input", String.valueOf(task.input(0)));
```
Comment (AGulev) [inline com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/Project.java:1795]: It's too much data for a big project. And this particular data isn't that useful

**Review by JCash — APPROVED** [review]

**Pull request merged**